Add intro and revise the layout (#9)
wangshan/draft-wang-ppm-differential-privacy: draft-wang-ppm-differential-privacy.md

# Introduction

Multi-party computation systems like the Distributed Aggregration Protocol
{{!DAP=I-D.draft-ietf-ppm-dap-05}} enable secure aggregation of measurements
generated by individuals without handling the measurements in the clear. This
is made possible by using a Verifiable Distributed Aggregation Function

> {{!VDAF=I-D.draft-irtf-cfrg-vdaf-06}}, the core cryptographic component of DAp.

Reviewer: ```suggestion
{{!VDAF=I-D.draft-irtf-cfrg-vdaf-06}}, the core cryptographic component of DAP.
```

Execution of A VDAF involves: a large set of "Clients" who produce
cryptographically protected measurements, called "reports"; a small number of
"Aggregators" who consume reports and produce the cryptographically protected
aggregate; and a "Collector" who consumes the aggregate result. Distributing
the computation of the aggregate in this manner ensures that, as long as one
Aggregator is honest, no attacker can learn an honest Client's measurement.

Depending on the application, protecting the measurements may not be sufficient
for privacy, since the aggregate itself can reveal privacy-sensitive
information. As an illustrative example, consider using DAP/VDAF to summarize
the distribution of the heights of respondents to a survey. If one of the
respondents is especially short or tall, then their contribution is likely to
skew the summary statistic in a way that reveals their height. Ideally, no
individual measurement would have such a signficant impact on the aggregate
result, but in general such leakage is inevitable.

> This intuition can be formalized by the notion differential privacy {{Dwo06}}.

Author: Is this a good reference or would you guys prefer something else? I don't have a strong opinion here, I just picked this one because it originally formalized DP.

Reviewer: I think either {{Dwo06}} or {{DMNS06}} is good.

Author: Changed to {{DMNS06}} so that we don't have more references than we need.

Differentially privacy is a property of an algorithm or protocol that computes
some function of a set of measurements. We say the algorithm or protocol is
"differentially private", or "DP", if the probability of observing a particular
output does not change signficantly as a result of removing one of the
measurements (or substituting it with another).

VDAFs are not DP on their own, but they can be composed with a variety of
mechanisms that endow them with this property. All such mechanisms work by
introducing "noise" into the computation that is carefully calibrated for a
number of application-specific parameters, including the structure and number
of measurements and the desired aggregation function. [TODO: Anything else?]

Noise can be introduced at various steps at the computation, and by various
parties. Depending on the mechanism: the Clients might add noise to their own
measurements; the Aggregators might add noise to their aggregate shares (the
values they produce for the Collector); and the Collector might add noise to
the aggregate result before publishing it. We refer to these classes of
mechanisms as "Client-DP", "Aggregator-DP", and "Collector-DP" respectively.

> [TODO: I put "Collector-DP" here as a strawman. Would we actually want to use

Reviewer: Don't think we will add collector-DP for now, because Collector would have unsharded the aggregate shares, and it defeats the purpose of secure aggregation. We can leave it in TODO though.

Author: Right, it's maybe worth mentioning as a possibility, but of course we would like VDAF execution to be DP from the point of view of the Collector as well. The straw man might be useful, so I'll leave it.

it?]

In addition, it is possible to compuse different DP mechanisms. [TODO: Say why
we would want to do this and give an illustrative example of combining
Client-DP and Aggregator-DP.]

[TODO: Transition text] This document defines a "DP policy" as the composition
of an optional Client-DP mechanism and an optional Aggregator-DP mechanism

> with a VDAF.

Author: @junyechen1996 this is how I suggested we define "DP policy" in https://github.com/wangshan/draft-wang-ppm-differential-privacy/pull/5#pullrequestreview-1534034920.

in DAP. It does so in the following stages:

1. {{overview}} provides an overview of DP properties in the literature that
   apply to VDAFs in general and DAP/VDAF in particular. Many refinements of
   the basic DP notion are possible, some of which are not possible given the
   constraints imposed by DAP. This section enumerates these constraints and
   describes the baseline requirements for a DP mechanism used with DAP.

1. {{primitives}} specifies various primitives required for implementing DP
   mechansims, including algorithms for sampling from discrete Laplace and
   Gaussian distributions.

1. {{mechanisms}} defines an interface for DP mechanisms that are compatible
   with VDAFs and, for all VDAFs specified in {{!VDAF}} [TODO: And other
   drafts, once they appear], specifies concrete Client- and Aggregator-DP
   mechanisms.

1. {{policy}} specifies the integration of DP mechanisms from the previous
   section into DAP. In particular, it describes how a DP policy is implemented
   by the Client, Aggregator, and Collector alongside the execution of the
   VDAF.

The following considerations are out-of-scope for this document:

> 1. Apart from privacy, the primarly consideratiaon for choosing a DP policy and

Reviewer: ```suggestion
1. Apart from privacy, the primary consideration for choosing a DP policy and
```

   calibrating it for an application is "utility". Intuitively, the more noise
   is added, the stronger the DP guarantee, but adding too much noise can

>    reduce the varacity of the result. This document provides no guidance for

Reviewer: ```suggestion
   reduce the veracity of the result. This document provides no guidance for
```

   this selection process.

1. This document describes a particular composition of narrowly-scoped DP
   mechanisms. Other compositions and more sophisticated DP mechanisms are
   possible.

> 1. The primitives described in {{primitives}} are intended for use beyond

Reviewer: I see your point. The current #4 probably fits under {{primitives}} better. It doesn't mean to be coupled with a specific VDAF yet. I can rebase once you merge this one.

Author: I'm proposing to split the current {{dp mechanisms}} section: I'm thinking the `DpMechanism` interface would fit into `{{mechanism}}` and we'd fit the low-level algorithms for sampling things in {{primitives}}. Does that sound good?

> TODO: Chris P: we will mention Prio3SumVec because that's what we use to describe aggregator DP with amplification

# Primitives for Implementing DP Mechanisms {#primitives}

This section describes various primitives required for implementing DP
mechansims. The algorithms are designed to securely expand a short, uniform
random seed into a sample from a given distribution.

## Discrete Laplace

> > TODO: Specify a Laplace sampler

Reviewer: For discrete Laplace and Gaussian, we can probably point people to [{{CKS20}}](https://arxiv.org/abs/2004.00010). Do we need to specify any PRG requirement for the sampler as well? e.g. how to derive seed, etc.

Author: I think so, but we could probably reuse the PRGs defined in the VDAF draft.

The purpose is to have a reference implementation of these algorithms so that we increase the likelihood of others developing secure implementations. (My understanding is that it's easy to mess up here in a way that invalidates DP.) If the algorithm spells out expanding a seed into a sample from the given distribution, then we can produce test vectors to help increase confidence.

Could you file an issue for specifying algorithms {{CKS20}}? No reason to actually do the work between now and Monday, but it'll be good to have our work scoped out.

@tholop may be able to help out here.

## Discrete Gaussian

> TODO: Specify a Gaussian sampler

## Randomized Response

> > TODO: Specify any primitives required for randomized response mechanisms

Author: @junyechen1996 what are examples of these?

Reviewer: The earliest paper is probably the one from Google: https://arxiv.org/abs/1407.6981, we haven't cited that one here. [This paper](https://arxiv.org/abs/2211.10082) ({{MJTBp22}}) also talks about two instantiations, one for replacement-DP setting in Appendix A, and one for deletion-DP setting in Appendix C.

Author: In your opinion, do any of these algorithms need to be specified? The primary motivation for standardizing Laplace/Gaussian is that there are lots of ways to get the implementation wrong.

I'm familiar with RAPPOR and I'm not sure any algorithm there needs to be standardized.

Reviewer: I think it's worth spelling out discrete Gaussian and Laplace. Both are pretty popular options in Aggregator-DP setting.

For RAPPOR, I think having one as an example for Client-DP setting would be useful in the draft.

Author: OK, SGTM!

# DP Mechanisms for VDAFs {#mechanisms}

> The section defines a generic interface for DP mechanisms VDAFs. Compatible

Reviewer: So in this part we would specify some interface the like `Vdaf.add_noise_to_agg_share` experimental API in libprio today?

Author: That's right :) It would also specify the interface for Client-DP mechanisms.

> mechnamsms include

Reviewer: This sentence is not finished?

Author: Good catch, I forgot to finish this bit. I just deleted the second sentence and the paragraph below.
